dll/win32/shell32/folders/CFSFolder.cpp

```diff
@@ -1626,3 +1627,152 @@ HRESULT WINAPI CFSFolder::CallBack(IShellFolder *psf, HWND hwndOwner, IDataObjec
+                        NULL, OPEN_EXISTING, 0, NULL);
+    if (hFile == INVALID_HANDLE_VALUE)
+        return NULL;
+
+    // get the file size
+    DWORD cbGlobal = GetFileSize(hFile, NULL);
+    if (cbGlobal == INVALID_FILE_SIZE)
+    {
+        CloseHandle(hFile);
+        return NULL;
+    }
+
+    // allocate
+    HGLOBAL hGlobal = GlobalAlloc(GMEM_MOVEABLE, cbGlobal);
+    if (hGlobal == NULL)
+    {
+        CloseHandle(hFile);
+        return NULL;
+    }
+
+    // read it
+    LPVOID pvGlobal = GlobalLock(hGlobal);
+    DWORD cbRead;
+    if (!pvGlobal || !ReadFile(hFile, pvGlobal, cbGlobal, &cbRead, NULL) ||
+        cbRead != cbGlobal)
+    {
+        GlobalUnlock(hGlobal);
+        GlobalFree(hGlobal);
+        CloseHandle(hFile);
+        return NULL;
+    }
+
+    GlobalUnlock(hGlobal);
+
+    // close the file
+    CloseHandle(hFile);
+
+    // load the picture
+
+    HBITMAP hbm = NULL;
+    IPicture *pPicture = NULL;
+    IStream *pStream = NULL;
+    if (CreateStreamOnHGlobal(hGlobal, TRUE, &pStream) == S_OK)
+    {
+        OleLoadPicture(pStream, cbGlobal, FALSE, IID_IPicture, (LPVOID *)&pPicture);
+
+        // get the bitmap handle
+        if (pPicture)
+        {
+            pPicture->get_Handle((OLE_HANDLE *)&hbm);
+
+            // copy the bitmap handle
+            hbm = (HBITMAP)CopyImage(hbm, IMAGE_BITMAP, 0, 0, LR_CREATEDIBSECTION);
+
+            pPicture->Release();
+        }
+
+        pStream->Release();
+    }
+
+    GlobalFree(hGlobal);
+    return hbm;
+}
+
+HRESULT WINAPI CFSFolder::GetCustomViewInfo(ULONG unknown, SFVM_CUSTOMVIEWINFO_DATA *data)
+{
+    if (data == NULL)
+    {
+        return E_POINTER;
+    }
+    if (data->cbSize != sizeof(*data))
+    {
+        return E_FAIL;
+    }
+
+    data->hbmBack = NULL;
+    data->clrText = CLR_INVALID;
+
+    WCHAR szPath[MAX_PATH], szIniFile[MAX_PATH];
+    SHGetPathFromIDListW(pidlRoot, szPath);
+
+    // does the folder exists?
+    if (!PathIsDirectoryW(szPath))
+    {
+        return E_FAIL;
+    }
+
+    // build the ini file path
+    StringCchCopyW(szIniFile, _countof(szIniFile), szPath);
+    PathAppend(szIniFile, L"desktop.ini");
+
+    static LPCWSTR TheGUID = L"{BE098140-A513-11D0-A3A4-00C04FD706EC}";
+    static LPCWSTR Space = L" \t\n\r\f\v";
+
+	// get info from ini file
+    WCHAR szImage[MAX_PATH], szText[64];
+    if (PathFileExists(szIniFile) && !PathIsDirectoryW(szIniFile))
+    {
+        GetPrivateProfileStringW(TheGUID, L"IconArea_Image", L"", szImage, _countof(szImage), szIniFile);
+        GetPrivateProfileStringW(TheGUID, L"IconArea_Text", L"", szText, _countof(szText), szIniFile);
+
+        // load the image
+        if (szImage[0])
+        {
+            StrTrimW(szImage, Space);
+            if (PathIsRelativeW(szImage))
+            {
+                PathAppendW(szPath, szImage);
+                StringCchCopyW(szImage, _countof(szImage), szPath);
+            }
+            data->hbmBack = DoLoadPicture(szImage);
```

Comment:
Would it make sense to cache the bitmap, so that we don't access the disk every redraw?

Reply:
How can we determine whether the cache is old?

Reply:
I think the CFSFolder is replaced if the user browses to another folder,
so it would be reasonable to cache the bitmap for the lifetime of the `CFSFolder` object?

If the same object is re-used while browsing trough folders, then I would cache it until the user browses to another folder.

Reply:
Now cached. The image will be loaded at `CDefView::FillList`.